Add specification for `SegmentMax-16` (#28103)
openvinotoolkit/openvino: docs/articles_en/documentation/openvino-ir-format/operation-sets/operation-specs/arithmetic/segment-max-16.rst

> SegmentMax
> ===================

Reviewer: ```suggestion
SegmentMax
==========
```
Should number `=` be same as `heading` length?

Author: Fixed.

.. meta::
  :description: Learn about SegmentMax-16 - an arithmetic operation which computes the maximum values along segments of a tensor.

**Versioned name**: *SegmentMax-16*

**Category**: *Arithmetic*

**Short description**: *SegmentMax-16* operation finds the maximum value in each specified segment of the ``input`` tensor.

**Detailed description**

For each index in ``segment_ids`` the operator gets values from ``data`` input tensor and calculates the maximum value for each segment.

For example ``segments_ids`` with value ``[0,0,0,1,1,3,5,5]`` defines 4 non-empty segments. The other segments are empty. When coupled with a 1D data tensor ``data``, the segments are as follows:

* Segment_0: ``[data[0], data[1], data[2]]``
* Segment_1: ``[data[3], data[4]]``
* Segment_2: ``[]``
* Segment_3: ``[data[5]]``
* Segment_4: ``[]``
* Segment_5: ``[data[6], data[7]]``

> When there are no values in a segment, ``output[segment]`` is set to 0.

Reviewer 2: we should have default value for empty segments, otherwise, we will have additional computation graph (that is not trivial) to compute empty segments and replace zero value

Author: The default value seems to be 0, according to https://www.tensorflow.org/api_docs/python/tf/raw_ops/SegmentMax.

Reviewer 3: There is also V2 https://www.tensorflow.org/api_docs/python/tf/raw_ops/SegmentMaxV2 where the default has been changed to `numeric_limits<T>::lowest()`. Adding attribute for default value seems to be a simple solution to support both cases, but to enable V2 at once we would also need to consider "num_segments" input.

Author: I think we can consider this resolved, an attribute/input for empty segment value has been added.

In that case, the output would be ``[max(Segment_0), max(Segment_1), 0, max(Segment_3), 0, max(Segment_5)]``.

**Attributes**:

SegmentMax-16 has no attributes.

**Inputs**

* **1**: *data*

>   * **Description**: The numerical data on which SegmentMax operation will be performed. **Required.**

Reviewer 2: please define input shapes and output shape for each input and describe what dimensions are equal

  * **Range of values**: An ND tensor of type *T*.
  * **Type**: *T*

* **2**: *segment_ids*

  * **Description**: Controls how the data is divided into segments. **Required.**

>   * **Range of values**: 1D tensor of non-negative, sorted integer numbers. Its size is equal to the size of the first dimension of the input tensor.

Reviewer 3: I can see that unsorted `segment_ids` may lead to error or undefined behavior (implementation specific, depends on the hardware).
Should we specify a common behavior for OV op?
Can be clarified at the plugin implementation stage.

Author: I assumed that we'll verify the orderliness of `segment_ids` during shape inference and raise an error if needed. We can always remove that check from shape inference later and unify the behavior, but I suppose for now the op will be enabled in CPU with reusage of reference implementation.

>   * **Range of values**: 1D tensor of non-negative, sorted integer numbers. Its size is equal to the size of the first dimension of the input tensor.

Reviewer 3: ```suggestion
  * **Range of values**: 1D tensor of non-negative, sorted integer numbers. Its size is equal to the size of the first dimension of the ``data`` input tensor.
```

Author: Added.

  * **Type**: *T_IDX*

**Outputs**

> * **1**: The output tensor of type *T* and the same shape as the ``input`` tensor with the exception for the first dimension, which is equal to the count of unique segment IDs.

Reviewer 3: The style of the "Inputs" section description follows rather the "Attributes" style.
Consider alignment with other spec documents.

Author: Fixed.

> * **1**: The output tensor of type *T* and the same shape as the ``input`` tensor with the exception for the first dimension, which is equal to the count of unique segment IDs.

Reviewer: Maybe instead `almost` use,  
```
The output tensor has same rank and dimensions as the ``data`` input tensor except first dimension which is calculated as ``max(segment_ids) + 1`` 
```
?

SegmentMax
==========

.. meta::
  :description: Learn about SegmentMax-16 - an arithmetic operation which computes the maximum values along segments of a tensor.

**Versioned name**: *SegmentMax-16*

**Category**: *Arithmetic*

**Short description**: *SegmentMax-16* operation finds the maximum value in each specified segment of the ``input`` tensor.

**Detailed description**

For each index in ``segment_ids`` the operator gets values from ``data`` input tensor and calculates the maximum value for each segment.

For example ``segments_ids`` with value ``[0,0,0,1,1,3,5,5]`` defines 4 non-empty segments. The other segments are empty. When coupled with a 1D data tensor ``data``, the segments are as follows:

* Segment_0: ``[data[0], data[1], data[2]]``
* Segment_1: ``[data[3], data[4]]``
* Segment_2: ``[]``
* Segment_3: ``[data[5]]``
* Segment_4: ``[]``
* Segment_5: ``[data[6], data[7]]``

When there are no values in a segment, ``output[segment]`` is defined by ``empty_segment_value`` input.

> In that case, the output would be ``[max(Segment_0), max(Segment_1), 0, max(Segment_3), 0, max(Segment_5)]``.

Author: ```suggestion
For ``empty_segment_value`` equal to 0, the operator output would be ``[max(Segment_0), max(Segment_1), 0, max(Segment_3), 0, max(Segment_5)]``.
```

**Attributes**:

* **1**: *empty_segment_value*

  * **Description**: The value assigned to segments which are empty. **Required.**
  * **Range of values**: A scalar.

>   * **Type**: *T*

Reviewer 2: let us have it as an input. it will be more convenient if this value is not constant to compute.

Author: After some discussion with @mitruska I modified the `empty_segment_value` to still be an attribute, but now is called `fill_mode` and is of type `string`. The reason is that it's not an input to the TF operation, but a way to differentiate between [SegmentMax](https://www.tensorflow.org/api_docs/python/tf/math/segment_max) and [SegmentMaxV2](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SegmentMaxV2). It allows us to create one operator and enable both of these.

The reason why its type has been changed to `string` is that the value `LOWEST` inserts the lower limit of type `T` into empty segments, and that lower value should be calculated at runtime because we may not know the type during operator conversion and some precision transformations may break the op.

What do you think? Can it stay as an attribute in its current form?

**Inputs**

* **1**: ``data`` - ND tensor of type *T*, the numerical data on which SegmentMax operation will be performed. **Required.**

* **2**: ``segment_ids`` - 1D Tensor of sorted non-negative numbers of type *T_IDX*. Its size is equal to the size of the first dimension of the ``data`` input tensor. The values must be smaller than ``num_segments``. **Required.**

> * **4**: ``num_segments`` - A scalar value of type *T_IDX* representing the segments count, used for shape inference. **Optional.**

Author: ```suggestion
* **3**: ``num_segments`` - A scalar value of type *T_IDX* representing the segments count, used for shape inference. **Optional.**
```

> * **2**: ``segment_ids`` - 1D Tensor of sorted non-negative numbers of type *T_IDX*. Its size is equal to the size of the first dimension of the ``data`` input tensor. The values must be smaller than ``num_segments``. **Required.**
> 
> * **3**: ``num_segments`` - A scalar value of type *T_IDX* representing the segments count, used for shape inference. **Optional.**

Reviewer 2: ```suggestion
* **2**: ``segment_ids`` - 1D Tensor of sorted non-negative numbers of type *T_IDX1*. Its size is equal to the size of the first dimension of the ``data`` input tensor. The values must be smaller than ``num_segments``. **Required.**

* **3**: ``num_segments`` - A scalar value of type *T_IDX2* representing the segments count, used for shape inference. **Optional.**
```
they can be of different type: https://www.tensorflow.org/api_docs/python/tf/raw_ops/SegmentMaxV2

**Outputs**

* **1**: The output tensor has same rank and dimensions as the ``data`` input tensor except for the first dimension which is calculated as ``max(segment_ids) + 1`` 
**Types**

* *T*: any supported numerical data type.

> * *T_IDX*: ``int64`` or ``int32``.

Reviewer 2: ```suggestion
* *T_IDX1*, *T_IDX2*: ``int64`` or ``int32``.
```

SegmentMax
==========

.. meta::
  :description: Learn about SegmentMax-16 - an arithmetic operation which computes the maximum values along segments of a tensor.

**Versioned name**: *SegmentMax-16*

**Category**: *Arithmetic*

**Short description**: *SegmentMax-16* operation finds the maximum value in each specified segment of the ``input`` tensor.

**Detailed description**

For each index in ``segment_ids`` the operator gets values from ``data`` input tensor and calculates the maximum value for each segment.

For example ``segments_ids`` with value ``[0,0,0,1,1,3,5,5]`` defines 4 non-empty segments. When coupled with a 1D data tensor ``data``, the segments are as follows:

* Segment_0: ``[data[0], data[1], data[2]]``
* Segment_1: ``[data[3], data[4]]``
* Segment_2: ``[]``
* Segment_3: ``[data[5]]``
* Segment_4: ``[]``
* Segment_5: ``[data[6], data[7]]``

When there are no values in a segment, ``output[segment]`` is defined by ``fill_mode`` attribute.

For ``fill_mode`` equal to ``ZERO`` , the operation output would be ``[max(Segment_0), max(Segment_1), 0, max(Segment_3), 0, max(Segment_5)]``.

**Attributes**:

> * **1**: *empty_segment_value*

Reviewer 3: To align with description and code it should be updated as well.
```suggestion
* **1**: *fill_mode*
```

  * **Description**: The value assigned to segments which are empty. **Required.**
  * **Range of values**: Name of the mode in string format:

    * ``ZERO`` - the empty segments are filled with zeros.
    * ``LOWEST`` - the empty segments are filled with the lowest value of the data type *T*.

>   * **Type**: ``string``

Reviewer 2: Let it be a constant input (the fourth one) of type *T* please. It will be more flexible for future use cases.
Attribute is injected in IR and no way to change it in run-time

Author: Since one of the values is `LOWEST`, which means the lowest value of type `T`, wouldn't this require custom behavior for graph type conversions? Setting that value dynamically during runtime could be safer.

Reviewer 2: I don't know but I don't exclude that this operation can be used in some decompositions to express some new FW operation. If we make an attribute for it, we create additional constraints. I think we don't need addtional constraints and let it make more flexible.

**Inputs**

* **1**: ``data`` - ND tensor of type *T*, the numerical data on which SegmentMax operation will be performed. **Required.**

* **2**: ``segment_ids`` - 1D Tensor of sorted non-negative numbers of type *T_IDX1*. Its size is equal to the size of the first dimension of the ``data`` input tensor. The values must be smaller than ``num_segments``. **Required.**

> * **3**: ``num_segments`` - A scalar value of type *T_IDX2* representing the segments count, used for shape inference. **Optional.**

Reviewer 2: Since this is optional, please write its value if it is not specified. 
As I understand, it will be a maximum value in `segment_ids`

Reviewer 2: Also, we need to write smth what will be if `num_segments < max(segment_ids) + 1`

Author: Done, waiting with a push for the first comment to be resolved.

**Outputs**

> * **1**: The output tensor has same rank and dimensions as the ``data`` input tensor except for the first dimension which is calculated as ``max(segment_ids) + 1`` 

Reviewer 2: ``max(segment_ids) + 1`` is not true always. If `num_segments` is specified it will be equal to `num_segments`

Author: Fixed, waiting with a push for the first comment to be resolved.
